packages/editor/src/containers/GridData.ts

```diff
@@ -70,27 +98,45 @@ export class GridData extends EventEmitter {
 
         const oldX16 = this._x16
         const oldY16 = this._y16
-        this._x16 = Math.floor(this._x / 16)
-        this._y16 = Math.floor(this._y / 16)
 
         const oldX32 = this._x32
         const oldY32 = this._y32
-        this._x32 = Math.floor(this._x / 32)
-        this._y32 = Math.floor(this._y / 32)
 
-        if (this.bpc.mode === EditorMode.PAN) return
+        const x16 = Math.floor(this._x / 16)
+        const y16 = Math.floor(this._y / 16)
 
-        // emit update when mouse changes tile whithin the 1 pixel size grid
-        if (!(oldX === this._x && oldY === this._y)) {
-            this.emit('update', this._x, this._y)
-        }
-        // emit update16 when mouse changes tile whithin the 16 pixel size grid
-        if (!(oldX16 === this._x16 && oldY16 === this._y16)) {
-            this.emit('update16', this._x16, this._y16)
+        const x32 = Math.floor(this._x / 32)
+        const y32 = Math.floor(this._y / 32)
+
+        if (this._constrained && this._constrainTo === undefined) {
+            const deltaX = Math.abs(x16 - oldX16)
+            const deltaY = Math.abs(y16 - oldY16)
+            if (deltaX > 0 && deltaX > deltaY) {
+                this._constrainTo = 'x'
+            } else if (deltaY > 0) {
+                this._constrainTo = 'y'
+            }
         }
-        // emit update32 when mouse changes tile whithin the 32 pixel size grid
-        if (!(oldX32 === this._x32 && oldY32 === this._y32)) {
-            this.emit('update32', this._x32, this._y32)
+
+        let more = true
+        while (more) {
+            more = this.constrainMove('_x16',  x16) || this.constrainMove('_y16',  y16) ||
+                   this.constrainMove('_x32',  x32) || this.constrainMove('_y32',  y32)
+
+            if (this.bpc.mode === EditorMode.PAN) return
+
+            // emit update when mouse changes tile whithin the 1 pixel size grid
+            if (!(oldX === this._x && oldY === this._y)) {
+                this.emit('update', this._x, this._y)
+            }
```

Comment:
Can we move these lines out of the loop?

Reply:
Yes good idea

Reply:
The pan mode check should also go out (before emitting any events).

Reply:
It needs to go after the update, though. I could repeat the update inside the pan check if you like that better.

Reply:
It should be fine actually since while panning all the values will (should) remain the same.

Reply:
The check can even be moved all the way up where we check if we have a `this.bpc`

Reply:
Ok, done